pkg/controller/machine.go

```diff
@@ -404,13 +404,24 @@ func (c *Controller) syncHandler(key string) error {
 					return fmt.Errorf("failed to create bootstrap kubeconfig: %v", err)
 				}
 
-				data, err := userdataProvider.UserData(machine.Spec, kubeconfig, prov, c.clusterDNSIPs)
+				//Need the kubeconfig to extract the CACert for the userdata rendering
+				cm, err := c.kubeconfigProvider.GetKubeconfig()
 				if err != nil {
-					return fmt.Errorf("failed get userdata: %v", err)
+					return fmt.Errorf("failed to get kubeconfig: '%v'", err)
+				}
+				if len(cm.Clusters) != 1 {
+					return fmt.Errorf("kubeconfig does not contain exactly one cluster, can not extract cacert...")
+				}
+				var userData string
+				for _, clusterConfig := range cm.Clusters {
+					userData, err = userdataProvider.UserData(machine.Spec, kubeconfig, prov, c.clusterDNSIPs, string(clusterConfig.CertificateAuthorityData))
```

Comment:
Maybe passing the kubeconfigProvider into the `UserData` function would make more sense.
As this is very specific logic in the main sync loop

Reply:
Because of performance reasons or because of readability? This only gets called when a new machine without associated instance at the cloud provider exists.

I dislike the idea of putting the whole kubeconfigProvider into `UserData`, because the latter only needs the CA from it.

I could move the whole `getCAAndUseItToRenderUserdata` into a seperate function within the `controller` package thought, that would probably improve readability.

Reply:
The UserData function needs the kubeconfig & ca data (depending if you agree with my suggestion above) from it.

If you don't agree, it's not needed